commcare_connect/opportunity/forms.py

```diff
@@ -287,8 +310,38 @@ def clean(self):
                 self.add_error("start_date", "Start date should be today or latter")
             if start_date >= end_date:
                 self.add_error("end_date", "End date must be after start date")
+
+            if self.opportunity.managed:
+                try:
+                    managed_opportunity = ManagedOpportunity.objects.get(id=self.opportunity.id)
+                    program = managed_opportunity.program
+                    if not (program.start_date <= start_date <= program.end_date):
+                        self.add_error("start_date", "Start date must be within the program's start and end dates.")
+
+                    if not (program.start_date <= end_date <= program.end_date):
+                        self.add_error("end_date", "End date must be within the program's start and end dates.")
+
+                    total_budget_sum = (
+                        ManagedOpportunity.objects.filter(program=program).aggregate(total=Sum("total_budget"))[
+                            "total"
+                        ]
+                        or 0
+                    )
+                    if total_budget_sum + cleaned_data["total_budget"] > program.budget:
+                        self.add_error("total_budget", "Budget exceeds the program budget.")
+                except ManagedOpportunity.DoesNotExist:
+                    # Should not happen.
+                    pass
+
             return cleaned_data
 
+    def save(self, commit=True):
```

Comment:
nit: This logic can be moved to the OpportunityFinalize view. Most of the object-saving-related logic is in the view.

Reply:
Changed